libs/contrib/Data/String/Parser.idr

```diff
@@ -85,81 +85,114 @@ parseT p str = do res <- p.runParser (S str 0 (strLength str))
-                            else pure $ Fail (s.pos) "satisfy"
-
-||| Succeeds if the string `str` follows.
-export
-string : Monad m => String -> ParseT m ()
-string str = P $ \s => do let len = strLength str
-                          if s.pos+len <= s.maxPos
-                              then do let head = strSubstr s.pos len s.input
-                                      if head == str
-                                        then pure $ OK () (S s.input (s.pos + len) s.maxPos)
-                                        else pure $ Fail (s.pos) ("string " ++ show str)
-                              else pure $ Fail (s.pos) ("string " ++ show str)
-
-||| Succeeds if the next char is `c`
+||| Fail with some error message
 export
-char : Monad m => Char -> ParseT m ()
-char c = do _ <- satisfy (== c)
-            pure ()
+total
+fail : Monad m => String -> ParseT m a
+fail x = P $ \s => pure $ Fail s.pos x
 
 mutual
     ||| Succeeds if `p` succeeds, will continue to match `p` until it fails
     ||| and accumulate the results in a list
     export
+    covering
     some : Monad m => ParseT m a -> ParseT m (List a)
     some p = pure (!p :: !(many p))
 
     ||| Always succeeds, will accumulate the results of `p` in a list until it fails.
     export
+    covering
     many : Monad m => ParseT m a -> ParseT m (List a)
     many p = some p <|> pure []
 
-||| Always succeeds, applies the predicate `f` on chars until it fails and creates a string
-||| from the results.
-export
-takeWhile : Monad m => (Char -> Bool) -> ParseT m String
-takeWhile f = do ls <- many (satisfy f)
-                 pure $ pack ls
-
 ||| Returns the result of the parser `p` or `def` if it fails.
 export
+total
 option : Monad m => a -> ParseT m a -> ParseT m a
 option def p = p <|> pure def
 
 ||| Returns a Maybe that contains the result of `p` if it succeeds or `Nothing` if it fails.
 export
+total
 optional : Monad m => ParseT m a -> ParseT m (Maybe a)
-optional p = (p >>= \res => pure $ Just res) <|> pure Nothing
+optional p = (Just <$> p) <|> pure Nothing
 
 ||| Discards the result of a parser
 export
+total
 skip : Parser a -> Parser ()
 skip = ignore
 
+||| Parse left-nested lists of the form `((init op arg) op arg) op arg`
+export
+covering
+hchainl : Parser init -> Parser (init -> arg -> init) -> Parser arg -> Parser init
+hchainl pini pop parg = pini >>= go
+  where
+  go : init -> Parser init
+  go x = (do op <- pop
+             arg <- parg
+             go $ op x arg) <|> pure x
+
+||| Parse right-nested lists of the form `arg op (arg op (arg op end))`
+export
+covering
+hchainr : Parser arg -> Parser (arg -> end -> end) -> Parser end -> Parser end
+hchainr parg pop pend = go id <*> pend
+  where
+  go : (end -> end) -> Parser (end -> end)
+  go f = (do arg <- parg
+             op <- pop
+             go $ f . op arg) <|> pure f
+
+||| Succeeds if the next char satisfies the predicate `f`
+export
+satisfy : Monad m => (Char -> Bool) -> ParseT m Char
```

Comment:
This ought to be `total` (even if that means you need to use an `assert`)
given that the bound checks are performed.